controllers/controller.go

```diff
@@ -224,23 +236,60 @@ func (i instanceReconcilerHelper) reconcileInstance(ctx context.Context, o clien
 	return ctrl.Result{}, nil
 }
 
-func (i instanceReconcilerHelper) checkPreconditions(o client.Object) (bool, ctrl.Result, error) {
-	i.rec.Event(o, corev1.EventTypeNormal, eventWaitingforPreconditions, "waiting for preconditions of the instance")
+func (i instanceReconcilerHelper) checkPreconditions(ctx context.Context, o client.Object) (bool, error) {
```

Comment:
What is the reason why we change this function signature? In this new implementation it handles error and re-queue  separately, so that a user cannot do both at once.

Reply:
Because of code duplication. Now there is two more returns with requeue now. 
It's also never checked if any error returned:

```
	if requeue, result, err := i.checkPreconditions(o); requeue {
		return result, err
	}

	if !isAlreadyProcessed(o) {
```

It either returns error or reschedules reconciliation.

Reply:
It was intentional not to check error, why - check this one https://github.com/aiven/aiven-operator/pull/206#discussion_r999436704

Reply:
That is a bit confusing. It's more common to check error for nil, rather than ignoring. Because when you extend checkPreconditions function then you have to know that it breaks the pattern. 
For instance you can send both requeue=false and err != nil and then it will pass. Error should not be ignored.